File: client-android/serenada-core/src/main/java/app/serenada/core/ForegroundMediaArbiter.kt
package app.serenada.core

import android.os.Looper
import java.util.Collections
import java.util.IdentityHashMap

/**
 * Owning mode of the process-wide foreground media arbiter (Core Invariant 6).
 *
 * A host integrates through direct single-call APIs OR the registry, not both at
 * once. The first user claims a mode; while that side has live sessions/calls the
 * other mode's acquire fails with [ForegroundLeaseUnavailable].
 */
enum class ForegroundArbiterMode {
    /** Claimed by the (Phase 3) registry; holds the process while it has any managed call. */
    REGISTRY,

    /** Claimed by a direct [SerenadaCore.join]; holds the process while a direct session is live. */
    DIRECT,
}

/**
 * Opaque foreground-lease owner token. Identity is the only thing that matters:
 * two tokens are equal iff they are the same object. Minted only by
 * [ForegroundMediaArbiter.acquireForeground]; a direct caller cannot forge one,
 * so a session cannot move itself to foreground without going through the arbiter.
 *
 * @property ownerId the call id (or session ref) the lease was acquired for; for diagnostics only.
 */
class ForegroundOwnerToken internal constructor(val ownerId: String)

/**
 * Thrown when the single foreground media lease cannot be granted: a lease is
 * already live, a prior owner's release is pending/failed, or the requested
 * owning mode conflicts with the mode currently owning the process.
 */
class ForegroundLeaseUnavailable(message: String) : Exception(message)

/**
 * Process-wide foreground media arbiter (multi-call session, contract §2 /
 * design "Process-Wide Resource Arbiter"). EXACTLY ONE instance per process:
 * the [ForegroundMediaArbiter] object below, shared by every [SerenadaCore] and
 * (Phase 3) every `SerenadaCallRegistry`. OS-global resources (mic, camera,
 * screen share, audio focus, `MODE_IN_COMMUNICATION`) are process-global, so a
 * per-core / per-registry arbiter would still let two owners race; this is the
 * single point that grants the one foreground lease.
 *
 * Concurrency: all callers operate on the main thread ([Looper.getMainLooper]),
 * the same thread the session's `Dispatchers.Main.immediate` scope and the
 * audio coordinator run on. State is therefore single-threaded and needs no
 * locks; [acquireForeground]/[releaseLease]/[claimMode]/[releaseMode] assert the
 * main thread. The async release-old-before-acquire-new sequencing lives in the
 * caller (Phase 2 single-call teardown; Phase 3 registry switch), not here; this
 * object is the synchronous bookkeeper of the lease + mode + generation.
 */
object ForegroundMediaArbiter {
    /** The live lease token, or null when no call owns foreground media. */
    private var currentToken: ForegroundOwnerToken? = null

    /**
     * The most recently released token, retained so an idempotent re-release of
     * the SAME token (after it was already released, nothing new granted) is a
     * safe no-op rather than a mismatch error.
     */
    private var lastReleasedToken: ForegroundOwnerToken? = null

    /**
     * True between starting an owner's release and confirming it fully drained.
     * While set (or after a failed release) NO new lease is granted — this is
     * what makes an old-release failure safe (never two owners; contract §2 rule
     * 2). Phase 2's single-call path releases synchronously on teardown and does
     * not set this; the Phase 3 registry switch uses [markReleasePending].
     */
    private var releasePending = false

    /** Monotonic operation generation; bumped per call. Separate from the token. */
    private var operationGeneration: Long = 0

    // --- Owning mode (Core Invariant 6) ---
    /** The mode currently owning the process, or null when no side has live calls. */
    private var mode: ForegroundArbiterMode? = null

    /**
     * Distinct owner references that have claimed the current mode. The mode
     * clears only when the last claimant releases. A `DIRECT` join claims with
     * its own session ref; the registry claims once with itself. Identity-keyed.
     */
    private val modeOwners = Collections.newSetFromMap(IdentityHashMap<Any, Boolean>())

    /**
     * Acquire the single foreground media lease for [ownerId]. The optional
     * [mode] claims/asserts the owning mode in the same step (common path: a
     * direct join acquires [ForegroundArbiterMode.DIRECT]; the Phase 3 registry's
     * foreground activation acquires [ForegroundArbiterMode.REGISTRY]). Throws
     * [ForegroundLeaseUnavailable] if a lease is already live, a prior release is
     * pending/failed, or the requested mode conflicts with the owning mode.
     */
    fun acquireForeground(
        ownerId: String,
        mode: ForegroundArbiterMode? = null,
        modeOwnerRef: Any? = null,
    ): ForegroundOwnerToken {
        assertMainThread()
        // Assert mode compatibility BEFORE granting so a cross-mode acquire fails
        // without side effects.
        if (mode != null) assertModeCompatible(mode)
        currentToken?.let { token ->
            throw ForegroundLeaseUnavailable(
                "Foreground lease already held by ${token.ownerId}; $ownerId cannot acquire",
            )
        }
        if (releasePending) {
            throw ForegroundLeaseUnavailable(
                "A prior foreground lease release is pending or failed; $ownerId cannot acquire",
            )
        }
        if (mode != null) claimMode(mode, modeOwnerRef ?: modeRefFor(ownerId))
        val token = ForegroundOwnerToken(ownerId)
        currentToken = token
        return token
    }

    /**
     * Release the lease held by [token]. Only the current owner's token is
     * accepted; a mismatched (non-current) token throws — EXCEPT the idempotent
     * case where [token] is the most recently released token and nothing new has
     * been granted, which is a safe no-op. Lease release is registry-owned (the
     * session never calls this); on the Phase 2 single-call path the session's
     * own teardown calls it.
     */
    fun releaseLease(token: ForegroundOwnerToken) {
        assertMainThread()
        if (currentToken === token) {
            currentToken = null
            releasePending = false
            lastReleasedToken = token
            return
        }
        // Idempotent re-release of the same token after it was already released.
        if (lastReleasedToken === token && currentToken == null) return
        throw ForegroundLeaseUnavailable(
            "releaseLease called with a token that is not the current owner (${token.ownerId})",
        )
    }

    /**
     * Mark that an owner's release has begun but not yet confirmed. Set by the
     * Phase 3 registry switch before it drains the old session; cleared by
     * [releaseLease] on success. While set, a new acquire fails fast so two
     * owners can never exist (contract §2 rule 2).
     */
    fun markReleasePending() {
        assertMainThread()
        releasePending = true
    }

    /** A monotonic operation generation, bumped every call. Not the lease identity. */
    fun nextOperationGeneration(): Long {
        assertMainThread()
        operationGeneration += 1
        return operationGeneration
    }

    /** Whether [token] is the live lease owner (used by the session's fence). */
    fun isCurrentOwner(token: ForegroundOwnerToken?): Boolean {
        return token != null && currentToken === token
    }

    /**
     * Claim the owning [mode] for [ownerRef] (Core Invariant 6). First user wins;
     * subsequent claims of the SAME mode just add the ref. A claim of the OTHER
     * mode while the current mode still has owners throws
     * [ForegroundLeaseUnavailable]. Folded into [acquireForeground]; exposed
     * separately so the Phase 3 registry can claim `REGISTRY` mode for held-only
     * calls (which take no lease).
     */
    fun claimMode(mode: ForegroundArbiterMode, ownerRef: Any) {
        assertMainThread()
        assertModeCompatible(mode)
        this.mode = mode
        modeOwners.add(ownerRef)
    }

    /**
     * Release [ownerRef]'s mode claim. The mode clears only when the last
     * claimant releases (the owning side has no live sessions/calls), after which
     * the other mode may claim it. A ref that never claimed is ignored.
     */
    fun releaseMode(ownerRef: Any) {
        assertMainThread()
        modeOwners.remove(ownerRef)
        if (modeOwners.isEmpty()) mode = null
    }

    /** The owning mode, or null when no side has live calls (diagnostic). */
    val currentMode: ForegroundArbiterMode?
        get() = mode

    private fun assertModeCompatible(mode: ForegroundArbiterMode) {
        val current = this.mode
        if (current != null && current != mode && modeOwners.isNotEmpty()) {
            throw ForegroundLeaseUnavailable(
                "Process is owned in '$current' mode; cannot acquire foreground media in '$mode' mode " +
                    "(direct single-call and registry-managed use cannot mix while either has live calls)",
            )
        }
    }

    // Stable per-ownerId mode-ref when the caller does not supply its own object,
    // so a single owner that claims a mode twice does not leave a dangling ref.
    private val modeRefs = HashMap<String, Any>()

    private fun modeRefFor(ownerId: String): Any = modeRefs.getOrPut(ownerId) { Any() }

    private fun assertMainThread() {
        check(Looper.myLooper() == Looper.getMainLooper()) {
            "ForegroundMediaArbiter must be used on the main thread"
        }
    }

    /**
     * Test-only: reset the process-global state so a lease/mode held by one test
     * cannot make a later sequential test fail to acquire. NOT part of the public
     * SDK surface — wire into the test harness setup/teardown
     * ([app.serenada.core.fakes.TestSessionFactory]). No-op for production code.
     */
    internal fun resetForTests() {
        currentToken = null
        lastReleasedToken = null
        releasePending = false
        operationGeneration = 0
        mode = null
        modeOwners.clear()
        modeRefs.clear()
    }
}

File: client-ios/SerenadaCore/Sources/ForegroundMediaArbiter.swift
import Foundation

/// Opaque owner token issued by ``ForegroundMediaArbiter`` to the holder of the
/// single process-wide foreground media lease. Two tokens are equal only when
/// they identify the same lease grant. The session uses it to fence stale async
/// activation callbacks; the registry uses it to release the lease.
public struct ForegroundOwnerToken: Equatable, Sendable {
    /// Process-unique identity of this lease grant (monotonic, never reused).
    fileprivate let id: Int
    /// The call/owner id this token was minted for (diagnostics only).
    public let ownerId: String

    fileprivate init(id: Int, ownerId: String) {
        self.id = id
        self.ownerId = ownerId
    }
}

/// Raised when a foreground lease cannot be acquired: a lease is already live, a
/// previous owner's release is still pending/failed, or the requested owning mode
/// conflicts with the mode currently in use (Core Invariants 1, 2, 6).
public struct ForegroundLeaseUnavailable: Error, Equatable, Sendable {
    /// Diagnostic reason for the failed acquisition.
    public enum Reason: String, Sendable {
        /// A lease is currently held by another owner.
        case leaseLive
        /// The previous owner's release has not yet been confirmed.
        case releasePending
        /// The other owning mode (registry vs direct) has live owners.
        case modeConflict
    }

    public let reason: Reason
    public init(reason: Reason) { self.reason = reason }
}

/// Which integration path owns the process-wide foreground media. Direct
/// single-call `SerenadaCore.join()` and the registry cannot be mixed while
/// either has live owners (Core Invariant 6).
public enum ForegroundOwningMode: String, Sendable {
    /// A `SerenadaCallRegistry` owns the process.
    case registry
    /// One or more direct `SerenadaCore.join()` sessions own the process.
    case direct
}

/// Process-wide arbiter granting at most one foreground media lease at a time.
///
/// Exactly one instance per process — `ForegroundMediaArbiter.shared`. Shared by
/// every `SerenadaCore` (direct single-call path) and every
/// `SerenadaCallRegistry`. OS audio session, mic, camera, and screen share are
/// process-global, so a per-core or per-registry arbiter would still let two
/// owners race (design "Add a Process-Wide Resource Arbiter").
///
/// The arbiter does NOT itself drive the audio coordinator or capture; it only
/// arbitrates the single lease + owning mode and vends operation generations for
/// fencing. The session+coordinator own the actual media transitions.
@MainActor
public final class ForegroundMediaArbiter {
    /// The single process-wide arbiter instance.
    public static let shared = ForegroundMediaArbiter()

    /// The live lease's token, or `nil` when no lease is held. A token is also
    /// retained here while a release is pending/failed (see ``releasePending``).
    private var currentToken: ForegroundOwnerToken?
    /// True between handing a token to a caller and that caller confirming the
    /// lease is released. While true, a new acquire fails (Core Invariant 2).
    /// In v1 the iOS release path is synchronous from the arbiter's point of
    /// view (the registry calls `releaseLease` only after the session confirms
    /// fully-held), so this is reset inside `releaseLease`; it exists as the
    /// explicit guard the contract requires.
    private var releasePending = false
    /// Monotonic id source for tokens (never reused, distinct from generation).
    private var nextTokenId = 1
    /// Monotonic operation-generation counter, bumped per activation attempt
    /// (including rollback). Separate from the token identity.
    private var operationGeneration = 0

    /// The owning mode claimed on first use, or `nil` when no side has live
    /// owners. Tracked separately from the lease: a registry holding only held
    /// calls (no foreground lease) still owns the process.
    private(set) var owningMode: ForegroundOwningMode?
    /// Set of live owner refs per mode. A direct join registers itself here on
    /// acquire and clears on release; the registry registers its managed-call
    /// presence (Phase 3). Mode clears when this drops to empty.
    private var liveOwnerRefs: Set<ObjectIdentifier> = []

    /// Public for the diagnostics/tests; the singleton is the only intended
    /// instance, but a fresh arbiter is useful for isolating unit tests.
    public init() {}

    // MARK: - Owning mode (Core Invariant 6)

    /// Claim the process owning mode for `ownerRef`. The first user wins; while
    /// the owning side has any live owner, claiming the other mode throws
    /// `ForegroundLeaseUnavailable(.modeConflict)`. Idempotent for an already
    /// registered ref in the same mode.
    func claimMode(_ mode: ForegroundOwningMode, ownerRef: AnyObject) throws {
        if let owningMode, owningMode != mode, !liveOwnerRefs.isEmpty {
            throw ForegroundLeaseUnavailable(reason: .modeConflict)
        }
        owningMode = mode
        liveOwnerRefs.insert(ObjectIdentifier(ownerRef))
    }

    /// Release `ownerRef`'s mode registration. When the owning side has no live
    /// owners left, the mode clears and the other mode may claim it.
    func releaseMode(ownerRef: AnyObject) {
        liveOwnerRefs.remove(ObjectIdentifier(ownerRef))
        if liveOwnerRefs.isEmpty {
            owningMode = nil
        }
    }

    // MARK: - Foreground lease

    /// Grant the single foreground lease to `ownerId`, returning a unique owner
    /// token. Throws `ForegroundLeaseUnavailable` if a lease is already live, a
    /// prior release is still pending/failed, or the caller is in a conflicting
    /// owning mode (Core Invariants 1, 2, 6).
    @discardableResult
    func acquireForeground(ownerId: String) throws -> ForegroundOwnerToken {
        if releasePending {
            throw ForegroundLeaseUnavailable(reason: .releasePending)
        }
        if currentToken != nil {
            throw ForegroundLeaseUnavailable(reason: .leaseLive)
        }
        let token = ForegroundOwnerToken(id: nextTokenId, ownerId: ownerId)
        nextTokenId += 1
        currentToken = token
        return token
    }

    /// Release the lease held under `token`. Only the current owner's token is
    /// accepted; a mismatched token is a programming error. Releasing the same
    /// token twice is idempotent (a no-op after the first release).
    func releaseLease(_ token: ForegroundOwnerToken) {
        guard let current = currentToken else {
            // Idempotent: already released. Tolerate a duplicate release of the
            // last-known token; reject an unrelated token.
            return
        }
        precondition(current == token, "releaseLease called with a non-owner token")
        currentToken = nil
        releasePending = false
    }

    /// Vend the next monotonic operation generation. Bumped for every activation
    /// attempt (including rollback) and passed into `activateForeground` so a
    /// superseded switch's late async callback can be discarded. NOT the lease
    /// identity.
    func nextOperationGeneration() -> Int {
        operationGeneration += 1
        return operationGeneration
    }

    /// The token of the live lease, or `nil`. Used by the session to fence late
    /// activation callbacks against the CURRENT owner (the second fence beyond
    /// the operation generation).
    var currentOwnerToken: ForegroundOwnerToken? { currentToken }

    // MARK: - Test support

    /// Reset all arbiter state. **Test-only** — the process singleton would
    /// otherwise leak a live lease/mode across XCTest cases and make a later
    /// `acquireForeground` fail. Call from test setUp/tearDown.
    func resetForTests() {
        currentToken = nil
        releasePending = false
        nextTokenId = 1
        operationGeneration = 0
        owningMode = nil
        liveOwnerRefs.removeAll()
    }
}
